src/What4/Serialize/Parser.hs

```diff
@@ -719,7 +757,71 @@ readApp rator rands = E.throwError $ ("readApp could not parse the following: "
 intToNatM :: (E.MonadError String m) => Integer -> m (Some NatRepr)
 intToNatM = fromMaybeError "integer must be non-negative to be a nat" . someNat
 
-
+-- | Parse a list of array updates where each entry in the list is:
+--
+-- > (idxs, elt)
+--
+-- where each @idxs@ is a list (assignment) of indexes (with type @idxReprs@)
+-- and each element is an expr.
+--
+-- NOTE: We assume that there are no duplicates in the list and apply the
+-- updates in an arbitrary order.  This is true for any map serialized by this
+-- library.
+expectArrayUpdateMap
+  :: forall sym t st fs tp i itp
+   . (sym ~ W4.ExprBuilder t st fs)
+  => Ctx.Assignment BaseTypeRepr (i Ctx.::> itp)
+  -> BaseTypeRepr tp
+  -> SExpr
+  -> Processor sym (WAU.ArrayUpdateMap (W4.SymExpr sym) (i Ctx.::> itp) tp)
+expectArrayUpdateMap idxReprs arrTyRepr updateSExprList =
+  case updateSExprList of
+    S.L items -> F.foldrM expectArrayUpdateEntry WAU.empty items
+    _ -> E.throwError "Expected a list of array element updates in ArrayMap"
+  where
+    expectArrayUpdateEntry pair updateMap =
+      case pair of
+        S.L [S.L idxListExprs, elt] -> do
+          idxs <- Ctx.traverseWithIndex (parseIndexLit idxListExprs) idxReprs
+          Some x <- readExpr elt
+          case testEquality arrTyRepr (W4.exprType x) of
+            Just Refl -> return (WAU.insert arrTyRepr idxs x updateMap)
+            Nothing -> E.throwError (concat [ "Invalid element type in ArrayMap update: expected "
+                                            , show arrTyRepr
+                                            , " but got "
+                                            , show (W4.exprType x)])
+        _ -> E.throwError "Unexpected ArrayMap update item structure"
+
+-- | Safe list indexing
+--
+-- This version only traverses the list once (compared to computing the length
+-- and then using unsafe indexing)
+(!?) :: [a] -> Int -> Maybe a
+lst !? idx
+  | idx < 0 = Nothing
+  | otherwise = go idx lst
+  where
```

Comment:
I'm wondering if some strictness annotations might be useful here.
I normally am not too concerned with those, but it seems like this
recursive evaluation could leave behind some lengthy thunk chains.